### PR TITLE
extend acceptance test to check for code style

### PR DESCRIPTION
**Pull request by mstreuhofer**

### CODE REVIEW

**Comment by mstreuhofer** [comment]
as seen here: https://travis-ci.org/kvz/bash3boilerplate/jobs/185239924#L1594 this is really really poor. the shellcheck people refuse to add this feature even. that's also why this check for style is running as part of the acceptance test but does not cause it to fail.

**Comment by zbeekman** [comment]
> as seen here: https://travis-ci.org/kvz/bash3boilerplate/jobs/185239924#L1594 this is really really poor. the shellcheck people refuse to add this feature even.

Sorry, I'm trying to understand what you're saying here but struggling. 

Are you saying that your style checking is poor and that you wish shellcheck had some support for  this? I think your implementation is really cool! Just one minor problem: Make sure to fail the build when there is a style violation: https://travis-ci.org/kvz/bash3boilerplate/jobs/185239924#L1612

File: test/acceptance.sh [hunk 1]
```diff
@@ -223,4 +224,39 @@ if [[ "$(command -v shellcheck)" ]]; then
...
+  echo "✓"
+done <<< "$(find . -type f -iname '*.sh')"
+
+popd > /dev/null
+
+if [[ "${failed}" = "true" ]]; then
+  echo
+  cat "${__accptstTmpDir}/style.err"
+  echo
+  exit 0
```
Comment (zbeekman) [inline test/acceptance.sh]: Don't we want to exit non-zero here?

**Comment by mstreuhofer** [comment]
yeah, sorry. what i mean is that my code for checking style (kind of) works but not really as it is nothing else but some fancy egrep. As you can see in the linked travis run it complains about an unbraces variable within an awk and sed script which of course it shouldn't complain about.

that is also reason why (on purpose) it does not fail (exit non-zero) the acceptance test. it wouldn't be fair to fail the tests just because my style check doesn't know the difference between bash and awk/sed.

For now i pushed that code so we can see the messages it would create yet not have any failed test just because of style. it's more like a "fyi..."

**Comment by kvz** [comment]
Nice initiative. It's too bad we can't get this approach to ever be fully correct. OTH, it helped you catch a few style violations so I'm inclined to keep it, while maybe adding a whitelist of exceptions so we can still make it fatal on errors. The purist in me has some issues with that, but the pragmatist is telling me it could help the project be more consistent, even if it's not perfect.

Perhaps both those of them can be pleased though. Dug up some more about this, it seems shellcheck is not configurable so we indeed cannot use that for our purposes if the authors aren't open to more opinionated style enforcing one way or the other. Perhaps we should turn to the node ecosystem and check out projects like:

- https://www.npmjs.com/package/bash-parser
- https://www.npmjs.com/package/falafel-bash (appears to use bash-parser under the hood)
- https://github.com/grncdr/js-shell-parse
- https://www.npmjs.com/package/bash-ast-traverser

Correct me if I'm wrong but I think I'm the only one with some Node.js experience so I can take a look at this. I'd probably aim low and implement just one check. And then maybe others can add more as the foundation is there. Still quite busy but perhaps there's some time over the holidays to look into this 🎄

**Comment by kvz** [comment]
In the meantime, I'd be okay merging this non-fatal check. It would be nice if we could add a `@todo`, referencing to this issue though

File: test/acceptance.sh (hunk 1 above)
Comment (zbeekman) [inline test/acceptance.sh]: Ah, I see. @mstreuhofer pointed out that the tests are not polished enough and create false positives...

Comment (zbeekman) [inline test/acceptance.sh]: Also, previous test failure will call `exit n` with `n > 0` so the exit 0 isn't an issue here.

**Comment by zbeekman** [comment]
> It would be nice if we could add a `@todo`, referencing to this issue though

What do you mean? A `@todo` in the source code? I'm categorically against `@todo` or `# FIXME` etc. in the source code because it gets lost and they're not all in one place, and then sometimes someone fixes the issue but doesn't delete the comment. Or they change the code and aren't sure whether the issue still applies and they just leave the comment in. Much prefer something like a GH issue. Just my $0.02

**Comment by mstreuhofer** [comment]
that egrep approach is way too simple either way so i wrote a less simple per script `test/style.pl` which tries to be smarter about it.

there should be no false positives but certainly some false negatives (since it removes everything between two single quotes).

`
echo "var='$var'"
`

will be changed into

`
echo "var='___'"
`

before checking for style which means that it won't complain about the unbraced `$var`.
still better than what we had before.

**Comment by mstreuhofer** [comment]
style check now tests for unbraced variables (`$VAR` vs `${VAR}`), single square brackets (`[` vs `[[`) and double equal sign (`[[ "${VAR} == "content" ]]` vs `[[ "${VAR} = "content" ]]`)

again, there is no guarantee that issues won't be overlooked.
there should be no false positives though.

**Comment by kvz** [comment]
Thanks for the work put into this @mstreuhofer! If we can agree this is a temporary solution until we can use AST (which is on me and might take a while to complete) I'm +1.

(note that I'm quite unfamiliar with perl, but the script seems small enough, and to be more about regexes than perl, that I'll likely get by 👍 😄 )

**Comment by kvz** [comment]
> What do you mean? A @todo in the source code? I'm categorically against @todo or # FIXME etc. in the source code because it gets lost and they're not all in one place, and then sometimes someone fixes the issue but doesn't delete the comment. Or they change the code and aren't sure whether the issue still applies and they just leave the comment in. Much prefer something like a GH issue. Just my $0.02

I disagree on this one. While I regard many `@todo`s in a project definitely as a code smell, they still have a valid purpose imho. There's also [plenty](https://github.com/imdone/imdone-atom) of tools [available](https://atom.io/packages/todo-show) to present all `@todo`s of a project in one overview . It's not a black/white thing to me. For most cases, gh issues are better. In this case, adding the todo also serves to signal people reading our code, that this is not the end game. They might otherwise walk away, or attempt to write fixes that we already have (different) plans for. Combined with the todo overview, this signals to our future selves that we had plans for improvement in that particular area. It's also not either/or to me: if discussion is warranted (pretty much always), you add the pointer to the thread. You can't expect just any reader of the code to also have exhaustive knowledge of our past & present github issues.

Sorry, didn't mean to derail this PR, but I think it's a useful discussion. In addition, I have a thing for flamewars 🔥😄 (I don't)

**Comment by zbeekman** [comment]
> Sorry, didn't mean to derail this PR, but I think it's a useful discussion. In addition, I have a thing for flamewars 🔥😄 (I don't)

No worries! We come from very different backgrounds, and you certainly have a more "professional developer" background than I do, so I value your perspective deeply. You're comment caused me to re-evaluate my own thinking on the matter and I appreciate that!

**Comment by mstreuhofer** [comment]
alright. this PR is now cleaned up and ready for merging.
i'll create a new GH-issue once it is merged linking here and requesting this homebaked "solution" of a style check be replaced with something better.
until then we will be reminded of its shortcomings anyway because i'm sure sooner or later some false positives will prevent a successful acceptance test.

**Pull request merged**